src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
```

Comment:
This optimization could remove a trap if the destination and source would lead to an out of bounds access. @kripken, is removing traps allowed here?

Reply:
@kripken friendly ping! This PR is ready to go except for this last question.

Reply:
Sorry for the delay!

In general, no - removing a trap is an observable thing we don't want to remove. (We do allow reordering of traps, though.)

However this could still be done if ignoreImplictTraps is set, as then we assume it won't trap.